src/biogeophys/HillslopeHydrologyMod.F90

```diff
@@ -0,0 +1,1326 @@
+       endif
+    enddo
+
+  end subroutine HillslopeSetLowlandUplandPfts
+
+  !------------------------------------------------------------------------
+  subroutine HillslopeDominantPftIndex(weights,lbound,pindex)
+    !
+    ! !DESCRIPTION:
+    ! Locate each gridcell's most dominant pft on the input dataset.
+    ! This is different than using n_dom_pts = 1, because it is meant
+    ! to be applied only to gridcells having active hillslope hydrology.
+    ! This routine is called from surfrd_hillslope.
+
+    !
+    ! !USES
+    !
+    ! !ARGUMENTS:
+    real(r8), intent(in)  :: weights(:)        ! array of patch weights
+    integer,  intent(in)  :: lbound            ! lower bound of weights
+    integer,  intent(out) :: pindex            ! index of largest weight
+    !
+    ! !LOCAL VARIABLES:
+    integer :: pdom(1)
+
+    !------------------------------------------------------------------------
+
+    pdom = maxloc(weights)
+    pindex = pdom(1) + (lbound - 1)
+
+  end subroutine HillslopeDominantPftIndex
+
+  !------------------------------------------------------------------------
+  subroutine HillslopeTwoLargestPftIndices(weights,lbound,pindex1,pindex2)
+    !
+    ! !DESCRIPTION:
+    ! Locate each gridcell's two most dominant patches on the input dataset.
+    ! This is different than using n_dom_pts = 2, because it is meant
+    ! to be applied only to gridcells having active hillslope hydrology.
+    ! This routine is called from surfrd_hillslope.
+
+    !
+    ! !USES
+    !
+    ! !ARGUMENTS:
+    real(r8), intent(in)  :: weights(:)        ! array of patch weights
+    integer,  intent(in)  :: lbound            ! lower bound of weights
+    integer,  intent(out) :: pindex1           ! index of largest weight
+    integer,  intent(out) :: pindex2           ! index of next largest weight
+    !
+    ! !LOCAL VARIABLES:
+    integer :: pdom(1),psubdom(1)
+    real(r8),allocatable :: mask(:)
+
+    !------------------------------------------------------------------------
+
+    pdom = maxloc(weights)
+    ! create mask to exclude pdom
+    allocate(mask(size(weights)))
+    mask(:) = 1.
+    mask(pdom(1)) = 0.
+    ! check that more than one nonzero patch weight exists,
+    ! if not return identical indices
+    if (sum(mask*weights) > 0) then
+       psubdom = maxloc(mask*weights)
+    else
+       psubdom = pdom
+    endif
+    deallocate(mask)
+
+    pindex1 = pdom(1) + (lbound - 1)
+    pindex2 = psubdom(1) + (lbound - 1)
+
+  end subroutine HillslopeTwoLargestPftIndices
+
+  !------------------------------------------------------------------------
+  subroutine HillslopeDominantPft(bounds)
+    !
+    ! !DESCRIPTION: 
+    ! Reassign patch weights of each column based on each gridcell's
+    ! most dominant pft on the input dataset. 
+    ! Assumes each column has a single pft.
+    ! If n_dom_pfts = 1 or if HillslopeDominantPftIndex is called
+    ! in surfrd_hillslope, this routine does not act on hillslope columns.
```

Comment:
- [ ] As you note in this comment, it seems like this routine no longer is necessary since you have already done the work in `surfrd_hillslope` for this case. Is that true? If so, should it be removed? (I see that there is a check for `col%npatches(c) > 1`, but that could be true even if the `surfrd_hillslope` stuff has already been done, e.g., in a transient run where we allocate memory for even zero-weight PFTs. Is there a reason to keep this routine at all?)